core/src/main/java/com/linecorp/armeria/client/endpoint/RingHashEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,205 @@
+                return null;
+            }
+            return weightedRingEndpoint.select(ctx.endpoint());
+        }
+        
+        private final class WeightedRingEndpoint {
+            private final Int2ObjectSortedMap<Endpoint> ring = new Int2ObjectAVLTreeMap<>();
+            private final Iterable<Endpoint> endpoints;
+
+            Endpoint select(Endpoint point) {
+                final int key = getXXHash(point.host());
+                final SortedMap<Integer, Endpoint> tailMap = ring.tailMap(key);
+                return tailMap.isEmpty() ? ring.get(ring.firstKey()) : tailMap.get(tailMap.firstKey());
+            }
+
+            WeightedRingEndpoint(Iterable<Endpoint> endpoints) {
+                // prepare immutable endpoints
+                this.endpoints = Streams.stream(endpoints)
+                                        .filter(e -> e.weight() > 0) // only process endpoint with weight > 0
+                                        .sorted(Comparator.comparing(Endpoint::weight)
+                                                          .thenComparing(Endpoint::host)
+                                                          .thenComparingInt(Endpoint::port))
+                                        .collect(toImmutableList());
+
+                // The GCD is properly sized so that it does not exceed the size of the ring
+                final int gcd = findGcdInEndpoints(this.endpoints);
+                final int numberOfEndpointInTheRing = caculateNumberOfEndpointInTheRing(this.endpoints, gcd);
+                final int sizeOfRing = getSize(this.endpoints);
+                if (sizeOfRing >= numberOfEndpointInTheRing) {
+                    for (Endpoint endpoint : this.endpoints) {
+                        final int weight = endpoint.weight();
+                        final String host = endpoint.host();
+                        // If weight is 3 and gcd is 1, place 3 times in the ring
+                        // if weight is 1 and gcd is 1, place once in the ring
+                        // If weight is 3 and gcd is 3, place 1 times in the ring
+                        final int count = weight / gcd;
+                        for (int i = 0; i < count; i++) {
+                            final String weightedHost = host + weight;
+                            final int hash = getXXHash(weightedHost);
+                            ring.put(hash, endpoint);
+                        }
+                    }
+                }
+
+                // When the size of the GCD is too small and exceeds the size of the ring,
+                // using binary search for find x where Σ (w[i] / x) ≤ ring_size, w[i] is endpoint's weight at index i
+                List<Integer> arr = new ArrayList<>();
+                for (Endpoint endpoint : this.endpoints) {
+                    int weight = endpoint.weight();
+                    arr.add(weight);
+                }
+                int x = binarySearch(arr, sizeOfRing);
```

Comment:
Could we give a more meaningful name to the method and returned value, for example?
```suggestion
                int divider = findClosestDivisor(arr, sizeOfRing);
```
```